### DISCUSSION TITLE
Content Blocks or "Page Builder" field

### DISCUSSION THREAD

**Discussion by FullStackAlex** [directus/directus#8310] [Feature Requests]
Hey, 
I am a WordPress (and Vue.js) developer and like their new Gutenberg editor pretty much. Unfortunately with its (Gutenberg's) introduction WordPress.org seems to have started targeting primarily end users without caring too much (if at all) about the feelings / interests of ("know-code") developers. Therefore I am looking for a new CMS platform which fits my needs (headless, great admin UX/UI & developer friendly). Directus seems to fit pretty well those. But there are still some things I am not sure about. Like for example the lack of an advanced content builder via a drag and drop editor like Gutenberg / Elementor / Brizzy or like Webflow for my clients. 

Are there any plans to introduce such page builders in addition to text and WYSIWYG fields? 
Or at least a field which can contain multiple content blocks (text or WYSIWYG sub fields) whose order can be rearranged via drag and drop and which can be created from within each single post / article editor?
Alternatively it would be great too if one just could extend the available fields with a page builder field based on another open source library (e.g. GrapesJS). But as I understand it's not possible as the extensions api is not released yet - is it right? 

Thanx in advance

**Comment by HCl-not-HCi** [comment]
You can use the Many-to-Any relation type (https://docs.directus.io/concepts/relationships/#many-to-any-m2a) for this.
- Create a collection for each content type (e.g. WYSIWYG or Image) you need (you have to "design" each content type yourself)
- In your Articles collection, create a new M2A Relationship field
- In the field setup under Relationship, you can link to your created content type collections

Alternatively, you can use [this](https://github.com/dimitrov-adrian/directus-extension-editorjs-interface) extension which implements editor.js.

There is also a discussion about implementing a block editor [here](https://github.com/directus/directus/discussions/5776).

**Reply by hermann-el2** [reply]
Well, I think I did exactly this
- Created a collection for a simple block containing a (sub-)title, textblock and image
- Created the M2A relationship field
- linked the textblock which I created
When I look at the article in the UI it looks fairly OK:
![image](https://user-images.githubusercontent.com/60488701/137389285-0c867143-95ec-42dd-8311-7aedb8838a0d.png)
when I click on each of the items, the correct textblock appears in the editor.

However, when I retrieve the data for this article (I get the same result for REST and GraphQL), the response looks like this:
```
{
    "data": {
        "id": "83bcc638-822b-433c-b971-769d32f6b548",
        "status": "draft",
        "sort": null,
        "user_created": "e8ce3ef8-6f49-4d64-ad5d-95951b26dddd",
        "date_created": "2021-10-14T17:12:25.289Z",
        "user_updated": "e8ce3ef8-6f49-4d64-ad5d-95951b26dddd",
        "date_updated": "2021-10-14T19:07:54.409Z",
        "date_published": "2022-06-14T00:00:07",
        "title": "Howto feed your cat",
        "featured_image": "3152130c-a695-41b9-bb06-9fc11502a38d",
        "body_text": "<p>Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.</p>",
        "textblock": [
            1,
            2
        ]
    }
}
```

and the "textblock" element just contains an index, but no ID or content which appears useful to me. Would you have any idea what I can try to improve this?

Thanks in advance.

**Reply by HCl-not-HCi** [reply]
You have to exactly specify which relational fields you want to query. This means, in GraphQL, your query should look something like this. Don't forget the `item` field which comes from the (in the app usually invisible) juction table connecting the textblock field and the items in the M2A collections you assigned to the field:

```graphql
{
  articles {
    id
    status
    title
    textblock {
      item {
        __typename
        ... on article_textblock {
          id
          text
        }
      }
    }
  }
}
```

This is also described here for REST: https://docs.directus.io/reference/api/query/#many-to-any-union-types
and here in general for GraphQL: https://graphql.org/learn/schema/#union-types

**Comment by FullStackAlex** [comment]
Cool, thank you! Sounds as a good starting point.  I'll try both ideas out.
